### PR TITLE
WIP: Support case-sensitive package loading

### PR DESCRIPTION
**Pull request by laszlocsomor**
DO NOT SUBMIT: PackageIdentifier just became case-sensitive; must figure out how to put that behind a flag.

The --incompatible_validate_package_path_casing
flag enables a new validation step in package
loading: when enabled, Bazel requires that package
paths be case-correct.

When the user runs "bazel build //Foo/Bar:Baz",
Bazel looks up the package "Foo/Bar" by checking
"Foo/Bar/BUILD". If the file exists then the
package exists. Every Label that Bazel creates in
this package is derived from the "Foo/Bar" string.

On Linux the filesystem is case-sensitive, so the
path casing must be correct for stat() to succeed.
It's impossible to load the package "//Foo/Bar" as
"//FOO/baR".

On Windows and macOS the filesystem ignores the
difference between lower and upper case letters,
so stat("foo/bAr/BUILD") and stat("FOO/bar/build")
both succeed, and "bazel build //foo/bAr:Baz" and
"bazel build //FOO/bar:Baz" both work. This leads
to bugs such as #6401, #8009, and #8799.

This PR adds a new SkyFunction
(PathCasingLookupFunction) that compares a
RootedPath's casing to the actual casing on disk.
The SkyKey for PathCasingLookupFunction stores not
just the RootedPath reference but also the String
representation, to distinguish differntly cased
but otherwise equal RootedPaths.

The PackageLookupFunction depends on the
PathCasingLookupValue for the path in the
PackageIdentifier, and denies loading the package
if the path was incorrectly cased.

Change-Id: I9d55cd76b2499172cb55252032163d936b83690e

### CODE REVIEW

**Comment by laszlocsomor** [comment]
This PR still needs:
- passing the CI tests
- adding a RELNOTES[NEW] tag
- creating a bug for the incompatible flag

**Comment by laszlocsomor** [comment]
@ulfjack: PTAL at the description.

File: src/main/java/com/google/devtools/build/lib/cmdline/PackageIdentifier.java [hunk 1]
```diff
@@ -116,7 +119,9 @@ public static PackageIdentifier discoverFromExecPath(PathFragment execPath, bool
   private PackageIdentifier(RepositoryName repository, PathFragment pkgName) {
     this.repository = Preconditions.checkNotNull(repository);
     this.pkgName = Preconditions.checkNotNull(pkgName);
-    this.hashCode = Objects.hash(repository, pkgName);
+    this.repositoryCasing = repository.getName();
```
Comment (ulfjack) [inline src/main/java/com/google/devtools/build/lib/cmdline/PackageIdentifier.java]: I don't see why you need to make a copy here. The RepositoryName and pkgName aren't changing underneath you, are they?

**Comment by ulfjack** [comment]
In what cases can the RootedPath be different from the String? Why not just construct a new RootedPath with the correct casing?

File: src/main/java/com/google/devtools/build/lib/cmdline/PackageIdentifier.java (hunk 1 above)
Comment (laszlocsomor) [reply src/main/java/com/google/devtools/build/lib/cmdline/PackageIdentifier.java]: RepositoryName.equals is case-ignoring. If I only stored the RepositoryName reference, I couldn't tell apart PackageIdentifiers with differently cased RepositoryNames. So I need to also store the String representation.

**Comment by jin** [comment]
Closing this for now, please ping/reopen if you're intending to work on this.

**Pull request closed without merging**